packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx

```diff
@@ -87,6 +89,27 @@ function TabsComponent(props: Props): JSX.Element {
   }
 
   const {tabGroupChoices, setTabGroupChoices} = useTabGroupChoice();
+  // search params >
+  // local storage >
+  // specified defaultValue >
+  // first child with "default" attr >
+  // first tab item.
+  let defaultValue: string | null | undefined;
+  if (queryString) {
+    const searchKey =
+      typeof queryString === 'string' ? queryString : groupId || '';
+    defaultValue = new URLSearchParams(location.search).get(searchKey);
```

Comment:
that's duplicated logic. I'd like such logic to be extracted in a custom hook or something to avoid the duplication

What about:

```jsx
const tabQueryString = useTabQueryString()

tabQueryString.get();
tabQueryString.set(val);
```

You could also encapsulate the usage of router/history hooks there

Reply:
I prefer to sacrifice a little bit of duplication (single line in this case) in order to avoid unnecessary abstractions. However, I noticed you are in the other camp which I respect :) Refactored.